includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,424 @@
+
+		return $encoding;
+	}
+
+	/**
+	 * Find a given tag with a given attribute.
+	 *
+	 * If multiple tags match, this method will only return the first one.
+	 *
+	 * @param string $content   Content in which to find the tag.
+	 * @param string $element   Element of the tag.
+	 * @param string $attribute Attribute that the tag contains.
+	 * @return string|false The requested tag, or false if not found.
+	 */
+	private function find_tag( $content, $element, $attribute = null ) {
+		$matches = [];
+		$pattern = empty( $attribute )
+			? sprintf(
+				'/<%1$s[^>]*?>[^<]*(?:<\/%1$s>)?/i',
+				preg_quote( $element, '/' )
+			)
+			: sprintf(
+				'/<%1$s [^>]*?\s*%2$s=[^>]*?>[^<]*(?:<\/%1$s>)?/i',
+				preg_quote( $element, '/' ),
+				preg_quote( $attribute, '/' )
+			);
+
+		if ( preg_match( $pattern, $content, $matches ) ) {
+			return $matches[0];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Extract an attribute value from an HTML tag.
+	 *
+	 * @param string $tag       Tag from which to extract the attribute.
+	 * @param string $attribute Attribute of which to extract the value.
+	 * @return string|false Extracted attribute value, false if not found.
+	 */
+	private function extract_value( $tag, $attribute ) {
+		$matches = [];
+		$pattern = sprintf(
+			'/%s=(?:([\'"])(?<full>.*)?\1|(?<partial>[^ \'";]+))/',
+			preg_quote( $attribute, '/' )
+		);
+
+		if ( preg_match( $pattern, $tag, $matches ) ) {
+			return empty( $matches['full'] ) ? $matches['partial'] : $matches['full'];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Sanitize the encoding that was detected.
+	 *
+	 * If sanitization fails, it will return 'auto', letting the conversion
+	 * logic try to figure it out itself.
+	 *
+	 * @param string $encoding Encoding to sanitize.
+	 * @return string Sanitized encoding. Falls back to 'auto' on failure.
+	 */
+	private function sanitize_encoding( $encoding ) {
+		if ( ! function_exists( 'mb_list_encodings' ) ) {
+			return $encoding;
+		}
+
+		static $known_encodings = null;
+
+		if ( null === $known_encodings ) {
+			$known_encodings = array_map( 'strtolower', mb_list_encodings() );
+		}
+
+		if ( array_key_exists( strtolower( $encoding ), $this->encoding_map ) ) {
+			$encoding = $this->encoding_map[ strtolower( $encoding ) ];
+		}
+
+		if ( ! in_array( strtolower( $encoding ), $known_encodings, true ) ) {
+			return self::UNKNOWN_ENCODING;
+		}
+
+		return $encoding;
+	}
+
+	/**
+	 * Magic getter to implement lazily-created, cached properties for the document.
+	 *
+	 * @param string $name Name of the property to get.
+	 * @return mixed Value of the property, or null if unknown property was requested.
+	 */
+	public function __get( $name ) {
+		switch ( $name ) {
+			case 'xpath':
+				$this->xpath = new DOMXPath( $this );
+				return $this->xpath;
+			case 'head':
+				$this->head = $this->getElementsByTagName( 'head' )->item( 0 );
+				return $this->head;
+			case 'body':
+				$this->body = $this->getElementsByTagName( 'body' )->item( 0 );
```

Comment:
If `$this->head` or `$this->body` are null, should this create those elements just in time?

Reply:
It does. We're not defining the internal properties `$head` and `$body`, so the first time around, they don't exist and a call to `$this->head` will trigger the magic `__get()`. Within the magic getter, we dynamically set the property `$head`, so the next call will immediately retrieve it and not hit the magic getter anymore.